### PR TITLE
Vision Transformer for Alzheimer's Detection

### PR DESCRIPTION
**Pull request by daniel-cottrell**
Implemented an algorithm using a Vision Transformer to detect Alzheimers in brain scans, for the 2024 COMP3710 Final Project.

The file layout for this is
```plaintext
recognition/
└── CN_AD_Classification_46991656/
    ├── README.md
    ├── dataset.py
    ├── graphs/
    │   └── training_metrics_plot.png
    ├── modules.py
    ├── predict.py
    └── train.py
```
<br>
For more information, please refer to the README.md file.  By Daniel Cottrell, 46991656.

### CODE REVIEW

**Comment by prprakash02** [comment]
**This is an initial inspection**

**1) Recognition Problem:**

a) Algorithm solves the respective pattern recognition problem appropriately: **(+3)**
b) Algorithm implemented in PyTorch demonstrated and functions as intended to solve the respective problem: **(+2)**
c) Good design implemented in PyTorch that solves the problem: **(+1)**
d) Commenting: **(+1)**
e) Algorithm is normal or above difficulty while implemented in PyTorch while solving the respective problem: Hard level **(+5)**
f) Algorithm is hard difficulty while implemented in PyTorch while solving the respective problem: Hard level **(+5)**

**(Total: 17/20)**

**2) Commit Log:**

a) Meaningful commit messages with evidence of individual work and originality: **(+2)**
b) Progressive commits and logical log structure providing evidence of individual work and originality: **(+3)**

**(Total: 5/5)**

**3) Documentation:**

Requirement:

Tittle: Done
Description of the algorithm: Done
Problem that it solves: Done
How it works in a paragraph (example any pre-processing): Done
Figure/visualisation (any input, output & results): Done
Dependencies: Done

Marking:

a) Description and explanation of the working principles of the algorithm implemented and the problem it solves: **(+5)**
b) Description of usage and comments throughout scripts: **(+3)**
c) Proper formatting using GitHub markdown: **(+2)**

**(Total: 10/10)**

**4) Pull Request**

a) Creating a pull-request into correct branch with a working and demonstrable version of the algorithm: **(+2)**
b) Incorporating feedback into the pull request: **(+2)**
c) Description of and comments within the pull request: **(+1)**

**(Total: 5/5)**

**Feedback:**

- **The README mentions a GFNet based model but only has a simple ViT**
- Could have a better variational model suitable for this task
- Could have tuned hyper-param, along with appropriate data augmentation using validation set

**Any Changes Needed: No**

**Comment by daniel-cottrell** [comment]
Ok thanks! I was in a rush when doing the README file, the model was meant to say vision transformer. I made an extra commit to fix this :)

**Comment by daniel-cottrell** [comment]
Hey @prprakash02 ,

I ran my train.py file that I had previously, and I found that it was flatlining around 51.6%. I'm not sure why it wasn't working, but I changed some parameters for it just today and ran it, and it is working significantly better now. I didn't change the model structure, just the parameters. The most notable change was reducing the patch size from 32 to 16, which increased the upper training accuracy to 83.75%, but the test accuracy was only 63.3%. However the code runs correctly, and the train.py file and predict.py file work as expected using rangpur. I updated the README.md file as well with the outputs from the updated model run today.

Cheers,
Daniel

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request should include clear description about the file structure .
Previous feedback incorporated .

File Organizing: Well-organized files.

Commit Log:
Commit messages are  progressive for the Recognition Problem solved using 4 files .
Commits show a logical developement .

Documentation:
The README file should contain in depth explanation of the algorithm and the problem it solves.
Code comments and docstrings are included.
Proper GitHub markdown formatting is used,  but should have more  organised headings, lists, and code blocks.
Can add  Conclusion  in the ReadMe.

**Comment by daniel-cottrell** [comment]
Ok thanks, I have added the changes suggested to the original pull request comment and the README.md file.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Adequate design and implementation. </td><td>-1</td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. </td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. ViT (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. Referencing missing.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-5</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
